### PR TITLE
Statically define connection creation and provisioing

### PR DESCRIPTION
**Pull request by Half-Shot**
This aims to nearly remove all burden on developers to modify anything outside their connection class when adding new connections. When creating a connection, developers now define their provisioning and creation functions as static functions on their connection class.

Connections can be added by simply adding a `@Connection` decorator.

```ts
@Connection
export class YourConnection {
....
}
```

### CODE REVIEW

File: src/Connections/JiraProject.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
-import { IConnection, IConnectionState } from "./IConnection";
-import { Appservice } from "matrix-bot-sdk";
+import { Connection, IConnection, IConnectionState, InstantiateConnectionOpts, ProvisionConnectionOpts } from "./IConnection";
```
Comment (Half-Shot) [inline src/Connections/JiraProject.ts:1]: For simplicity while we evaluate this idea, I've not committed the changes I've made to other repos.

**Review by tadzik — APPROVED** [review]
Huge +1 from me :) Few nitpicks.

**Review by tadzik — CHANGES_REQUESTED** [review]
­

File: src/Bridge.ts [hunk 2]
```diff
@@ -515,7 +514,10 @@ export class Bridge {
                     instance = CLOUD_INSTANCE;
                 } else {
                     tokenInfo = await this.tokenStore.jiraOAuth.exchangeRequestForToken(msg.data.oauthToken, msg.data.oauthVerifier);
-                    instance = new URL(this.config.jira.url!).host;
+                    if (!this.config.jira.url) {
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply src/Bridge.ts:517]: Hm, yeah probably. A empty URL is valid, but the config object should probably set a default value in this case.

File: src/Bridge.ts [hunk 3]
```diff
@@ -831,14 +833,22 @@ export class Bridge {
                     break;
                 }
             }
-            if (!handled && this.config.checkPermissionAny(event.sender, BridgePermissionLevel.manageConnections)) {
+            if (!handled && this.connectionManager && this.config.checkPermissionAny(event.sender, BridgePermissionLevel.manageConnections)) {
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply src/Bridge.ts]: Not sure, removed.

File: src/Bridge.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Half-Shot) [reply src/Bridge.ts]: Have hopefully solved this now.

**Review by tadzik — APPROVED** [review]

**Pull request merged**